File: examples/mission_upload_download.rs
use mavkit::{
    CompareTolerance, GeoPoint3d, MissionItem, MissionPlan, MissionType, NavWaypoint, Vehicle,
    normalize_for_compare, plans_equivalent,
};

fn waypoint(lat: f64, lon: f64, alt: f32) -> MissionItem {
    NavWaypoint::from_point(GeoPoint3d::rel_home(lat, lon, f64::from(alt))).into()
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let bind_addr =
        std::env::var("MAVKIT_EXAMPLE_UDP_BIND").unwrap_or_else(|_| "0.0.0.0:14550".to_string());

    let vehicle = Vehicle::connect_udp(&bind_addr).await?;
    let mission = vehicle.mission();

    let plan = MissionPlan {
        mission_type: MissionType::Mission,
        items: vec![
            waypoint(47.397742, 8.545594, 25.0),
            waypoint(47.398100, 8.546100, 30.0),
        ],
    };

    mission.upload(plan.clone())?.wait().await?;
    let downloaded = mission.download()?.wait().await?;

    let expected = normalize_for_compare(&plan);
    let got = normalize_for_compare(&downloaded);

    println!(
        "roundtrip equivalent: {}",
        plans_equivalent(&expected, &got, CompareTolerance::default())
    );

    vehicle.disconnect().await?;
    Ok(())
}

File: mavkit-python/examples/mission_upload_download.py
import asyncio
import os

import mavkit


def waypoint(seq: int, lat: float, lon: float, alt: float) -> mavkit.MissionItem:
    return mavkit.MissionItem(
        seq=seq,
        command=mavkit.NavWaypoint(
            latitude_deg=lat,
            longitude_deg=lon,
            altitude_m=alt,
            frame=mavkit.MissionFrame.GlobalRelativeAltInt,
        ),
        current=seq == 0,
    )


async def main():
    bind_addr = os.environ.get("MAVKIT_EXAMPLE_UDP_BIND", "0.0.0.0:14550")

    vehicle = await mavkit.Vehicle.connect_udp(bind_addr)
    try:
        mission = vehicle.mission()

        plan = mavkit.MissionPlan(
            mission_type=mavkit.MissionType.Mission,
            items=[
                waypoint(0, 47.397742, 8.545594, 25.0),
                waypoint(1, 47.398100, 8.546100, 30.0),
            ],
        )

        issues = mavkit.validate_plan(plan)
        if issues:
            raise RuntimeError(f"mission plan has {len(issues)} validation issue(s)")

        upload_op = mission.upload(plan)
        await upload_op.wait()

        download_op = mission.download()
        downloaded = await download_op.wait()

        equivalent = mavkit.plans_equivalent(
            mavkit.normalize_for_compare(plan),
            mavkit.normalize_for_compare(downloaded),
            mavkit.CompareTolerance(),
        )

        upload_progress = upload_op.latest()
        download_progress = download_op.latest()

        print(f"mission handle: {mission}")
        print(
            f"upload phase: {upload_progress.phase if upload_progress else 'unknown'}"
        )
        print(
            f"download phase: {download_progress.phase if download_progress else 'unknown'}"
        )
        print(f"uploaded items: {len(plan.items)}")
        print(f"downloaded items: {len(downloaded.items)}")
        print(f"roundtrip equivalent: {equivalent}")
    finally:
        await vehicle.disconnect()


asyncio.run(main())
